### ISSUE TITLE
Very slow response times (postgres 12)

### ISSUE THREAD

**Issue by rogierlommers** [spotweb/spotweb#624]
I'm using spotweb (latest release) with postgres (12.4) and the response times are very slow. This was never the case; it started suddenly. Now I'm wondering: what is the best way to debug this? Please see screenshot below; nothing really happened and I can't figure out why response times are increasing.

<img width="998" alt="Screenshot 2020-11-07 at 11 21 39" src="https://user-images.githubusercontent.com/3907131/98438590-8e325a00-20eb-11eb-84dd-0bdde308adc8.png">

**Comment by mesa57** [comment]
You can activate timings reporting in the settings.

**Comment by rogierlommers** [comment]
> You can activate timings reporting in the settings.

Ah! I enabled that, now I see that the query below takes around 16 seconds. But my follow-up question would be: how to optimize this? Did I miss some db migrations which create indexes or something? I already ran the `update-db.php` file again. No improvement.

```
select x.*, f.verified , COALESCE(x.idtype, wl.idtype, gwl.idtype) AS idtype from (SELECT s.id AS id, s.messageid AS messageid, s.category AS category, s.poster AS poster, s.subcata AS subcata, s.subcatb AS subcatb, s.subcatc AS subcatc, s.subcatd AS subcatd, s.subcatz AS subcatz, s.title AS title, s.tag AS tag, s.stamp AS stamp, s.moderated AS moderated, s.filesize AS filesize, s.spotrating AS rating, s.commentcount AS commentcount, s.reportcount AS reportcount, s.spotterid AS spotterid, s.editstamp AS editstamp, s.editor AS editor, l.download as downloadstamp, l.watch as watchstamp, l.seen AS seenstamp, bl.idtype as idtype, s.reversestamp as reversestamp FROM spots AS s LEFT JOIN spotteridblacklist as bl ON ((bl.spotterid = s.spotterid) AND ((bl.ouruserid = 3) OR (bl.ouruserid = -1)) AND (bl.idtype = 1)) LEFT JOIN spotstatelist AS l on ((s.messageid = l.messageid) AND (l.ouruserid = 3)) WHERE (bl.spotterid IS NULL) ORDER BY s.reversestamp ASC LIMIT 251 OFFSET 0 ) as x LEFT JOIN spotsfull AS f ON (f.messageid = x.messageid) LEFT JOIN spotteridblacklist as wl on ((wl.spotterid = x.spotterid) AND ((wl.ouruserid = 3) AND (wl.idtype = 2))) LEFT JOIN spotteridblacklist as gwl on ((gwl.spotterid = x.spotterid) AND ((gwl.ouruserid = -1) AND (gwl.idtype = 2))) ORDER BY x.reversestamp ASC
```

**Comment by rogierlommers** [comment]
I think some index is missing.

<img width="767" alt="Screenshot 2020-11-07 at 13 39 32" src="https://user-images.githubusercontent.com/3907131/98441458-b5465700-20fe-11eb-958d-157d3a4704ce.png">

**Comment by Sweepr** [comment]
Weird, if it's missing then update-db.php would re-create that.

**Comment by rogierlommers** [comment]
> Weird, if it's missing then upgrade-db.php would re-create that.

Any idea how I can debug further?

**Comment by Sweepr** [comment]
> 
> 
> > Weird, if it's missing then upgrade-db.php would re-create that.
> 
> Any idea how I can debug further?

Does the index actually exist in your DB now? or is the entire table gone?

**Comment by mesa57** [comment]
Can you determine why Postgres says that an index would be helpfull ?
And what kind of index, on which columns ....

**Comment by rogierlommers** [comment]
The table has three indexes. Is this the correct number?

<img width="404" alt="Screenshot 2020-11-07 at 16 26 38" src="https://user-images.githubusercontent.com/3907131/98445090-0a419780-2116-11eb-9faf-060dffa2ebed.png">

**Comment by mesa57** [comment]
I only see 2 : 
![afbeelding](https://user-images.githubusercontent.com/6350469/98445338-838dba00-2117-11eb-9e13-84b2557de77d.png)

**Comment by mesa57** [comment]
Anyway, this query should work better, but will considerable time to change : 

```
select
	x.*,
        l.download as downloadstamp, l.watch as watchstamp, l.seen as seenstamp,
	f.verified ,
	coalesce(x.idtype, wl.idtype, gwl.idtype) as idtype
from
	(
	select
		s.id as id, s.messageid as messageid, s.category as category, s.poster as poster, s.subcata as subcata, s.subcatb as subcatb, s.subcatc as subcatc, s.subcatd as subcatd, s.subcatz as subcatz, s.title as title, s.tag as tag, s.stamp as stamp, s.moderated as moderated, s.filesize as filesize, s.spotrating as rating, s.commentcount as commentcount, s.reportcount as reportcount, s.spotterid as spotterid, s.editstamp as editstamp, s.editor as editor,  bl.idtype as idtype, s.reversestamp as reversestamp
	from
		spots as s
	left join spotteridblacklist as bl on
		((bl.spotterid = s.spotterid)
		and ((bl.ouruserid = 3)
		or (bl.ouruserid = -1))
		and (bl.idtype = 1))
	where
		(bl.spotterid is null)
	order by
		s.reversestamp asc
	limit 251 offset 0 ) as x
left join spotstatelist as l on
	((x.messageid = l.messageid)
 	and (l.ouruserid = 3))
left join spotsfull as f on
	(f.messageid = x.messageid)
left join spotteridblacklist as wl on
	((wl.spotterid = x.spotterid)
	and ((wl.ouruserid = 3)
	and (wl.idtype = 2)))
left join spotteridblacklist as gwl on
	((gwl.spotterid = x.spotterid)
	and ((gwl.ouruserid = -1)
	and (gwl.idtype = 2)))
order by
	x.reversestamp asc
```

**Comment by mesa57** [comment]
Please test this change from the develop branch.

**Comment by rogierlommers** [comment]
Do you want/need me to do something? I'm currently running latest release-version.

**Comment by Sweepr** [comment]
@rogierlommers 

Pull the latest fix from the Develop branch made by Mesa57 and let us know if that fixes your issue.

Thanks.

**Comment by rogierlommers** [comment]
Ok, running latest `develop`-branch now. All looking good; page responses are fast; retrieving works. Haven't seen anything broke yet. What specific parts do you want me to test?

**Comment by mesa57** [comment]
I don't know which tracing tool you used in the first post, but it would be nice to see that one.
The query you posted is used to browse the list of spots, so it is very often used.

**Comment by rogierlommers** [comment]
Looking fine now.

![Uploading Screenshot 2020-11-08 at 17.26.56.png…]()

**Comment by mesa57** [comment]
Ok, I will close the issue then. Feel free to reopen if the problem occurs again.

**Comment by rogierlommers** [comment]
But will it be merged into release? (I prefer running a released version)

**Comment by mesa57** [comment]
I leave that so @Sweepr

**Comment by Sweepr** [comment]
@rogierlommers  Sorr, totally missed this comment.

Releases are only being made against the Master branch, if you are 100% certain this is working fine we could push and make a release for it.

Let me know and we can create a new release.

**Comment by rogierlommers** [comment]
I ran develop for a day and it looked fine. Not sure what quality control you normally have? I think the @mesa57 can best determine what the impact / risks would be.

**Comment by mesa57** [comment]
Impact is low because there are not many postgres users and this change would only affect postgres.

**Comment by Sweepr** [comment]
Ok, merged these with Master and created 1.4.9

https://github.com/spotweb/spotweb/releases/tag/1.4.9
